File: Examples/julia/basics.jl
struct Point
    x::Int
    y::Int
end

norm(p::Point) = p.x * p.x + p.y * p.y

function fib(n)
    if n < 2
        return n
    end
    return fib(n - 1) + fib(n - 2)
end

function classify(n)
    if n < 0
        "negative"
    elseif n == 0
        "zero"
    elseif n < 10
        "small"
    else
        "large"
    end
end

println("Hello, Julia!")

s = 0
for i in 1:10
    s += i
end
println("sum = ", s)
println(fib(15))

xs = [5, 2, 9, 1]
println(sort(xs))
println(map(x -> x * x, xs))
println(filter(x -> x > 2, xs))
println(sum(xs))
println(join(xs, "-"))

squares = [x^2 for x in 1:5]
println(squares)
evens = [x for x in 1:10 if x % 2 == 0]
println(evens)

p = Point(3, 4)
println(norm(p))
println(p)

d = Dict("a" => 1, "b" => 2)
println(d["a"] + d["b"])
println(haskey(d, "a"))

text = "Hello, World"
println(length(text))
println(uppercase(text))
println(split(text, ", "))

println(classify(-1), " ", classify(0), " ", classify(5), " ", classify(50))

println(7 / 2)
println(7 ÷ 2)
println(7 % 3)
println(2^10)

println(xs[1], " ", xs[end])

for (i, v) in enumerate(["a", "b"])
    println(i, ":", v)
end

name = "Julia"
println("Hello, $name! 1+2=$(1+2)")

File: Examples/zig/basics.zig
const std = @import("std");

const Point = struct {
    x: i32,
    y: i32,

    fn norm(self: Point) i32 {
        return self.x * self.x + self.y * self.y;
    }
};

fn fib(n: i32) i32 {
    if (n < 2) return n;
    return fib(n - 1) + fib(n - 2);
}

fn classify(n: i32) []const u8 {
    return switch (n) {
        0 => "zero",
        1...9 => "small",
        else => "large",
    };
}

pub fn main() void {
    std.debug.print("Hello, Zig!\n", .{});

    var sum: i32 = 0;
    var i: i32 = 1;
    while (i <= 10) : (i += 1) {
        sum += i;
    }
    std.debug.print("sum = {d}\n", .{sum});
    std.debug.print("{d}\n", .{fib(15)});

    const xs = [_]i32{ 5, 2, 9, 1 };
    for (xs) |x| {
        std.debug.print("{d} ", .{x});
    }
    std.debug.print("\n", .{});

    const p = Point{ .x = 3, .y = 4 };
    std.debug.print("{d}\n", .{p.norm()});

    std.debug.print("{s} {s} {s}\n", .{ classify(0), classify(5), classify(50) });

    var total: i32 = 0;
    for (xs) |x| {
        if (x > 2) total += x;
    }
    std.debug.print("{d}\n", .{total});

    std.debug.print("{d} {d}\n", .{ 7 / 2, 7 % 3 });
    std.debug.print("{s}\n", .{"hello" ++ " " ++ "world"});
}
